### PR TITLE
Address warnings with filenames ending in fullstop on Windows

### PR DESCRIPTION
**Pull request by BrianHenryIE**
@lipemat opened  PR #5947 to address an issue with filenames of premium plugins.

> Warning: copy(-.): Failed to open stream: Permission denied

The failure was occurring with:

```
FileCache::import( 'plugin/advanced-sidebar-menu-pro-9.5.7.', 'C:\Users\box\AppData\Local\Temp/664e5754330ad-TsJObX.tmp' )
```

Tests with those values passed on MacOS, but Microsoft's documentation says:

> Do not end a file or directory name with a space or a period.

https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions

The primary fix for this is to `rtrim()` the `.` during `FileCache::validate_key()`: https://github.com/wp-cli/wp-cli/commit/c38a3713eb14701682d473ec10457541309b8103, confirmed working: https://github.com/wp-cli/wp-cli/pull/5947#issuecomment-2125784637

Before realising the root issue, I had added a fix for the error message, https://github.com/wp-cli/wp-cli/commit/32ba58dbf4d68ed257c3f5b213702ad52bde1bac, to begin, I needed to add a test for the happy path: https://github.com/wp-cli/wp-cli/commit/6f0b0e98f8059c4693cc45f36844766f5e7a88d6, and I also added some comments and typehints I needed to debug.

### CODE REVIEW

**Comment by swissspidy** [comment]
Some of the failing unit tests are addressed by #5950

File: tests/test-file-cache.php [hunk 1]
```diff
@@ -102,4 +102,104 @@ public function test_export() {
...
+		$zip = new ZipArchive();
+		$zip->open( $fixture_filepath, ZIPARCHIVE::CREATE );
+		$zip->addFile( __FILE__ );
+		$zip->close();
+
+		$result = $cache->import( $key, $fixture_filepath );
+
+		// Assert file is imported.
+		self::assertTrue( $result );
+		self::assertFileExists( "{$cache_dir}/{$key}" );
```
Comment (swissspidy) [inline tests/test-file-cache.php:128]:
```suggestion
		$this->assertTrue( $result );
		$this->assertFileExists( "{$cache_dir}/{$key}" );
```

File: tests/test-file-cache.php [hunk 2]
```diff
@@ -102,4 +102,104 @@ public function test_export() {
...
+				$error = $errstr;
+			}
+		);
+
+		$result = $cache->import( $key, $fixture_filepath );
+
+		restore_error_handler();
+
+		self::assertNull( $error );
+		self::assertFalse( $result );
```
Comment (swissspidy) [inline tests/test-file-cache.php:171]:
```suggestion
		$this->assertNull( $error );
		$this->assertFalse( $result );
```

File: tests/test-file-cache.php [hunk 3]
```diff
@@ -102,4 +102,104 @@ public function test_export() {
...
+
+		$reflection = new ReflectionClass( $cache );
+
+		$method = $reflection->getMethod( 'validate_key' );
+		$method->setAccessible( true );
+
+		$result = $method->invoke( $cache, $key );
+
+		self::assertFalse( str_ends_with( $result, '.' ) );
+		self::assertEquals( 'plugin/advanced-sidebar-menu-pro-9.5.7', $result );
```
Comment (swissspidy) [inline tests/test-file-cache.php:203]: `str_ends_with` does not exist on PHP 5.6+

```suggestion
		$this->assertStringEndsNotWith( '.', $result );
		$this->assertSame( 'plugin/advanced-sidebar-menu-pro-9.5.7', $result );
```

**Comment by BrianHenryIE** [comment]
Hmm... the PHPUnit methods [are static methods](https://github.com/sebastianbergmann/phpunit/blob/9.6.19/src/Framework/Assert.php#L330) but the [documentation's examples](https://docs.phpunit.de/en/9.6/assertions.html#assertequals) do use the instance `$this->`.

> A regular user of PHPUnit should use $this-> instead of self::

https://github.com/sebastianbergmann/phpunit/issues/1914#issuecomment-148407321

https://phpunit.de/manual/6.5/en/appendixes.assertions.html#appendixes.assertions.static-vs-non-static-usage-of-assertion-methods

**Review by swissspidy — APPROVED** [review]

**Pull request merged**